src/components/Experiments/ExperimentEditModal.js

```diff
@@ -0,0 +1,195 @@
+} from '../../state/experiments';
+
+import './ExperimentEditModal.less';
+
+class ExperimentEditModal extends Component {
+  constructor(props) {
+    super(props);
+
+    const exp = this.props.editModalExperiment;
+    if (exp !== undefined) {
+      this.state = { nameVal: exp.name, descriptionVal: exp.description };
+    } else {
+      this.state = {
+        nameVal: '',
+        descriptionVal: '',
+      };
+    }
+
+    this.handleClose = this.handleClose.bind(this);
+    this.handleNameChange = this.handleNameChange.bind(this);
+    this.handleDescriptionChange = this.handleDescriptionChange.bind(this);
+    this.handleSave = this.handleSave.bind(this);
+    this.handleUpload = this.handleUpload.bind(this);
+  }
+
+  handleClose() {
+    this.props.showEditModal(false);
+  }
+
+  handleNameChange(e) {
+    this.setState({ nameVal: e.target.value });
+  }
+
+  handleDescriptionChange(e) {
+    this.setState({ descriptionVal: e.target.value });
+  }
+
+  handleSave() {
+    const create = this.props.editModalExperiment === undefined;
+
+    if (create) {
+      this.props.experimentCreate(
+        this.state.nameVal,
+        this.state.descriptionVal
+      );
+    } else {
+      this.props.experimentUpdate(
+        this.props.editModalExperiment.id,
+        this.state.nameVal,
+        this.state.descriptionVal
+      );
+    }
+  }
+
+  handleUpload(e) {
+    this.props.uploadFilePairs(
+      this.props.editModalExperiment.id,
+      e.target.files[0]
+    );
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const exp = nextProps.editModalExperiment;
+    if (exp !== undefined) {
+      this.setState({ nameVal: exp.name, descriptionVal: exp.description });
+    } else {
+      this.setState({
+        nameVal: '',
+        descriptionVal: '',
+      });
+    }
+  }
+
+  render() {
+    const {
+      editModalShow,
+      createInProgress,
+      updateInProgress,
+      uploadInProgress,
+      uploadResult,
+    } = this.props.experiments;
+
+    const create = this.props.editModalExperiment === undefined;
+
+    const title = create ? 'Create Experiment' : 'Edit Experiment';
+    const btn = create ? 'Create' : 'Save';
+    const btnDisabled =
+      createInProgress || updateInProgress || this.state.nameVal.trim() === '';
+
+    const uploadDisabled = create || uploadInProgress;
+
+    let uploadIcon;
+    let uploadClass;
+    if (uploadInProgress) {
+      uploadIcon = 'floppy-open';
+      uploadClass = 'upload-floppy upload-floppy-progress';
+    } else if (uploadResult === UPLOAD_RES_SUCCESS) {
+      uploadIcon = 'floppy-saved';
+      uploadClass = 'upload-floppy upload-floppy-success';
+    } else if (uploadResult === UPLOAD_RES_FAILURE) {
+      uploadIcon = 'floppy-remove';
+      uploadClass = 'upload-floppy upload-floppy-failure';
+    } else {
+      uploadIcon = 'floppy-disk';
+      uploadClass = 'upload-floppy';
+    }
+
+    return (
+      <Modal show={editModalShow} onHide={this.handleClose}>
```

Comment:
I would recommend controlling `show` outside of the modal component. In general it's better to do `editModalShow  && <Modal>`. Otherwise it might not render virtual/real dom if the underlying component is written well, but all life cycle methods would be executed on every parent update.
Not very important in this particular case though.

Reply:
Makes sense. I decided to leave it as a property of the Modal component to avoid introducing more and more changes. Let me know if you think this should be changed considering the last changes.